locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/templates/customize/TemplateOptionsOption.java

```diff
@@ -39,8 +45,17 @@ public void apply(TemplateOptions options, ConfigBag config, Object v) {
 
         Class<? extends TemplateOptions> clazz = options.getClass();
         for (final Map.Entry<String, Object> option : optionsMap.entrySet()) {
-            if (option.getValue() != null) {
-                Maybe<?> result = MethodCoercions.tryFindAndInvokeBestMatchingMethod(options, option.getKey(), option.getValue());
+            Object optionValue = option.getValue();
+            if (optionValue != null) {
+
+                if (optionValue instanceof TaskFactory<?>) {
+                    final TaskAdaptable<?> taskA = ((TaskFactory<?>) optionValue).newTask();
+                    optionValue = getValueFromTask(taskA.asTask());
+                } else if (optionValue instanceof TaskAdaptable<?>) {
+                    optionValue = getValueFromTask(((TaskAdaptable<?>) optionValue).asTask());
+                }
+
+                Maybe<?> result = MethodCoercions.tryFindAndInvokeBestMatchingMethod(options, option.getKey(), optionValue);
```

Comment:
Don't think you can use `null` here. Seems the thread context is something used around immediate values only. cc @ahgittin.

Looking at the code the canonical usage is with: `((EntityInternal)BrooklynTaskTags.getTargetOrContextEntity(Tasks.current())).getExecutionContext()`.